packages/validators/typescript/src/TypeScriptValidator.js

```diff
@@ -48,18 +51,25 @@ export default new Validator({
         ts.sys,
         baseDir,
       );
-
-      langServiceCache[configHash] = ts.createLanguageService(
-        new LanguageServiceHost(options.inputFS, ts, parsedCommandLine),
-        ts.createDocumentRegistry(),
+      const host = new LanguageServiceHost(
+        options.inputFS,
+        ts,
+        parsedCommandLine,
       );
+      langServiceCache[configHash] = {
+        host,
+        service: ts.createLanguageService(host, ts.createDocumentRegistry()),
+      };
     }
 
     if (!langServiceCache[configHash]) return;
 
-    const diagnostics = langServiceCache[configHash].getSemanticDiagnostics(
-      asset.filePath,
-    );
+    // Make sure that when the typescript language service asks us for this file, we let it know that there is a new version.
+    langServiceCache[configHash].host.invalidate(asset.filePath);
```

Comment:
Wow, I thought typescript automatically did this. My bad

Reply:
Does Typescript consider more than one file at a time while type checking? Is it possible that multiple files may have changed, but we are only invalidating the one, which could lead to a false error?

Reply:
Yes, I do think that is possible, hopefully their invalidate function reruns validations over all imports as well. Not sure, would have to look up if there’s any documentation on it

Reply:
Based on my reading of the code (which may be wrong - please correct me if so), the `invalidate` function is not a typescript API - it's part of _our_ implementation of the `LanguageServiceHost` - see here:

https://github.com/parcel-bundler/parcel/blob/706a7e647f90c65724f6b3560376c9ffab80eb5e/packages/utils/ts-utils/src/LanguageServiceHost.js#L18-L28

 It has the effect of ensuring that the next time the Typescript Language Service calls getScriptVersion (which _is_ part of [the interface that a `LanguageServiceHost` must implement](https://github.com/microsoft/TypeScript/blob/7f02927b7803384b87c50c884b07cd872dbed1af/src/services/types.ts#L213)), it returns an incremented version number:

https://github.com/parcel-bundler/parcel/blob/706a7e647f90c65724f6b3560376c9ffab80eb5e/packages/utils/ts-utils/src/LanguageServiceHost.js#L34-L36

By setting breakpoints in my simple test app, I can see that that when you ask the Language Service for diagnostics about any given file (through `getSemanticDiagnostics`) it will make calls to `getScriptVersion` for every file that is a dependency of the file you are checking. Presumably, if we return an incremented version number, it can tell that it needs to re-analyze the dependency as well. 

I could see a possible bug here related to order (maybe that's what you were talking about above): if you have a project with `a.ts` that depends on `b.ts`, and you change both of them, we invalidate and check `a.ts` before we have invalidated `b.ts`, so you get a potentially false error on `a.ts` (which was analyzed assuming the old contents of `b.ts`. I'll test this scenario to confirm and think about ways to address it.

Reply:
I tested things out, and this is indeed a problem - I filed #4204 to track it.

Since it was a pre-existing problem (this PR merely unblocks you from discovering it), I'd propose merging this, and doing a separate PR to tackle that one.

Reply:
That's unfortunate, sounds good to fix this separate from this PR